File: Src/java/engine/src/test/java/org/opencds/cqf/cql/engine/elm/executing/PredecessorEvaluatorTest.java
package org.opencds.cqf.cql.engine.elm.executing;

import static org.junit.jupiter.api.Assertions.*;

import java.math.BigDecimal;
import org.junit.jupiter.api.Test;
import org.opencds.cqf.cql.engine.runtime.Quantity;

class PredecessorEvaluatorTest {

    @Test
    void predecessor() {
        assertEquals(
                new BigDecimal("19"),
                PredecessorEvaluator.predecessor(
                        new BigDecimal("20"), new Quantity().withValue(new BigDecimal("100"))));
        assertEquals(
                new BigDecimal("19.99"),
                PredecessorEvaluator.predecessor(
                        new BigDecimal("20"), new Quantity().withValue(new BigDecimal("100.00"))));

        var actualQuantity = (Quantity) PredecessorEvaluator.predecessor(
                new Quantity().withValue(new BigDecimal("20")).withUnit("g"),
                new Quantity().withValue(new BigDecimal("100.00")).withUnit("g"));
        assertEquals(new BigDecimal("19.99"), actualQuantity.getValue());
        assertEquals("g", actualQuantity.getUnit());
    }
}

File: Src/java/engine/src/jvmTest/kotlin/org/opencds/cqf/cql/engine/elm/executing/PredecessorEvaluatorTest.kt
package org.opencds.cqf.cql.engine.elm.executing

import java.math.BigDecimal
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.opencds.cqf.cql.engine.runtime.Quantity

internal class PredecessorEvaluatorTest {
    @Test
    fun predecessor() {
        Assertions.assertEquals(
            BigDecimal("19"),
            PredecessorEvaluator.predecessor(
                BigDecimal("20"),
                Quantity().withValue(BigDecimal("100")),
            ),
        )
        Assertions.assertEquals(
            BigDecimal("19.99"),
            PredecessorEvaluator.predecessor(
                BigDecimal("20"),
                Quantity().withValue(BigDecimal("100.00")),
            ),
        )

        val actualQuantity =
            PredecessorEvaluator.predecessor(
                Quantity().withValue(BigDecimal("20")).withUnit("g"),
                Quantity().withValue(BigDecimal("100.00")).withUnit("g"),
            ) as Quantity
        Assertions.assertEquals(BigDecimal("19.99"), actualQuantity.value)
        Assertions.assertEquals("g", actualQuantity.unit)
    }
}
